<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>dropletQC • dropletQC</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="dropletQC">
<meta property="og:description" content="dropletQC">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dropletQC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/dropletQC.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/WalterMuskovic/dropletQC/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="dropletQC_files/header-attrs-2.6/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>dropletQC</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/WalterMuskovic/dropletQC/blob/master/vignettes/dropletQC.Rmd"><code>vignettes/dropletQC.Rmd</code></a></small>
      <div class="hidden name"><code>dropletQC.Rmd</code></div>

    </div>

    
    
<div id="how-dropletqc-works" class="section level2">
<h2 class="hasAnchor">
<a href="#how-dropletqc-works" class="anchor"></a>How dropletQC works</h2>
<p>To calculate the nuclear fraction score for each provided cell barcode, we use the <code>nuclear_fraction_tags</code> function. This simply scans the ‘CB’ and ‘RE’ tags in the provided indexed Cell Ranger barcoded bam file. To access these tags it uses a strategy similar to this:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Rsamtools">Rsamtools</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges">GenomicRanges</a></span><span class="op">)</span>

<span class="co"># Create a reference to the BAM file</span>
<span class="va">bam_file</span> <span class="op">&lt;-</span>
   <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html">BamFile</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"outs"</span>, <span class="st">"possorted_genome_bam.bam"</span>,
                       package <span class="op">=</span> <span class="st">"dropletQC"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Define some genomic intervals to pull data from</span>
<span class="va">intervals</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="st">"chr1"</span>,
                     ranges <span class="op">=</span> <span class="fu">IRanges</span><span class="op">(</span>start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1180</span>, <span class="fl">4880</span><span class="op">)</span>,
                                      end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">1190</span>, <span class="fl">4890</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Pull CB and RE tags from the specified regions in the BAM file</span>
<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span>
   <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/scanBam.html">scanBam</a></span><span class="op">(</span><span class="va">bam_file</span>,
           param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/ScanBamParam-class.html">ScanBamParam</a></span><span class="op">(</span>
              tag <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CB"</span>, <span class="st">"RE"</span><span class="op">)</span>,
              which <span class="op">=</span> <span class="va">intervals</span><span class="op">[</span><span class="va">x</span><span class="op">]</span>
           <span class="op">)</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">tag</span><span class="op">)</span></code></pre></div>
<pre><code>##    [,1]                 [,2]                 [,3]                
## CB "AAAAGTCACTTACTTG-1" "AAAAGTCACTTACTTG-1" "AAAAGTCACTTACTTG-1"
## RE "E"                  "I"                  "N"</code></pre>
<p>‘CB’ is the error-corrected cell barcode sequence and ‘RE’ is one of three single characters representing the region type of the read; E (exonic), N (intronic) or I (intergenic). The nuclear fraction score that is returned by <code>nuclear_fraction_tags</code> is simply E/(E+N) for each provided cell barcode. To speed up the processing of the (usually large) BAM file, dropletQC splits the BAM file into a number of genomic tiles and processes them in parallel by internally calling <code><a href="https://rdrr.io/pkg/furrr/man/future_map.html">furrr::future_map</a></code>.</p>
</div>
<div id="calculating-the-nuclear-fraction-score" class="section level2">
<h2 class="hasAnchor">
<a href="#calculating-the-nuclear-fraction-score" class="anchor"></a>Calculating the nuclear fraction score</h2>
<p>There are two ways to use <code>nuclear_fraction_tags</code>; point it to the Cell Ranger ‘outs’ directory or point it to the individual files. The simplest way is to rely on the consistent structure of the ‘outs’ directory produced by the Cell Ranger software:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">dropletQC</span><span class="op">)</span>

<span class="va">nf1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/nuclear_fraction_tags.html">nuclear_fraction_tags</a></span><span class="op">(</span>
   outs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"outs"</span>, package <span class="op">=</span> <span class="st">"dropletQC"</span><span class="op">)</span>,
   tiles <span class="op">=</span> <span class="fl">10</span>,
   cores <span class="op">=</span> <span class="fl">1</span>,
   verbose <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">nf1</span><span class="op">)</span></code></pre></div>
<pre><code>##                    nuclear_fraction
## AAAAGTCACTTACTTG-1       0.05025126
## AAAAGTGGATCTCTAA-1       0.03804348
## AAACACGTTCTCATCG-1       0.02985075
## AAACAGGCAGCGACTG-1       0.06451613
## AAAGCAGTTACGAAGA-1       0.04624277
## AAAGCGGATGCATGGT-1       0.03821656</code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># This assumes the following three files are present in the specified directory:</span>
<span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"outs"</span>, package <span class="op">=</span> <span class="st">"dropletQC"</span><span class="op">)</span>,
           recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "filtered_feature_bc_matrix/barcodes.tsv.gz"
## [2] "possorted_genome_bam.bam"                  
## [3] "possorted_genome_bam.bam.bai"</code></pre>
<p>If you don’t have this directory structure, or your files have been renamed e.g. they were given to you by a collaborator, you can specify the paths to the required files directly:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">nf2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/nuclear_fraction_tags.html">nuclear_fraction_tags</a></span><span class="op">(</span>
   bam <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"outs"</span>, <span class="st">"possorted_genome_bam.bam"</span>, package <span class="op">=</span>
                        <span class="st">"dropletQC"</span><span class="op">)</span>,
   barcodes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>
      <span class="st">"AAAAGTCACTTACTTG-1"</span>,
      <span class="st">"AAAAGTGGATCTCTAA-1"</span>,
      <span class="st">"AAACACGTTCTCATCG-1"</span>
   <span class="op">)</span>,
   tiles <span class="op">=</span> <span class="fl">10</span>,
   cores <span class="op">=</span> <span class="fl">1</span>,
   verbose <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span>
<span class="va">nf2</span></code></pre></div>
<pre><code>##                    nuclear_fraction
## AAAAGTCACTTACTTG-1       0.05025126
## AAAAGTGGATCTCTAA-1       0.03804348
## AAACACGTTCTCATCG-1       0.02985075</code></pre>
<p>Note that here we have provided a vector of requested barcode IDs to the barcodes argument rather than the path to a file on disk ‘barcodes.tsv.gz’. Either is fine, just make sure the format of your barcodes matches the BAM file.</p>
</div>
<div id="assessing-the-nuclear-fraction-score" class="section level2">
<h2 class="hasAnchor">
<a href="#assessing-the-nuclear-fraction-score" class="anchor"></a>Assessing the nuclear fraction score</h2>
<p>Once you’ve calculated the nuclear fraction score you can use it to help identify damaged cells or empty droplets. We provide some pre-calculated scores, as well as some additional metrics and information, from four single cell RNA-seq datasets produced and made publicly available by 10x Genomics:</p>
<ol style="list-style-type: decimal">
<li><p>Mouse E18 Combined Cortex, Hippocampus and Subventricular Zone Cells, ~10k cells <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/4.0.0/SC3_v3_NextGem_DI_Neuron_10K" title="dataset link">dataset link</a></p></li>
<li><p>Human Glioblastoma Multiforme, ~5k cells <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/4.0.0/Parent_SC3v3_Human_Glioblastoma" title="dataset link">dataset link</a></p></li>
<li><p>PBMCs from a Healthy Donor, ~10k cells <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/4.0.0/Parent_NGSC3_DI_PBMC" title="dataset link">dataset link</a></p></li>
<li><p>Hodgkin’s Lymphoma, Dissociated Tumour, ~5k cells <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/4.0.0/Parent_NGSC3_DI_HodgkinsLymphoma" title="dataset link">dataset link</a></p></li>
</ol>
<p>Each dataset has been pre-filtered using <code><a href="https://rdrr.io/pkg/DropletUtils/man/emptyDrops.html">DropletUtils::emptyDrops</a></code> and a 15% mitochondrial gene content cutoff.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"qc_examples"</span><span class="op">)</span>

<span class="co"># What does the dataset contain?</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">qc_examples</span><span class="op">)</span></code></pre></div>
<pre><code>##                    sample       cell_barcode    umap_1     umap_2
## AAACCCAAGGCGATAC-1    GBM AAACCCAAGGCGATAC-1 -7.221113 -16.802812
## AAACCCAAGGCTGTAG-1    GBM AAACCCAAGGCTGTAG-1 -8.374323  -4.323620
## AAACCCACAAGTCCCG-1    GBM AAACCCACAAGTCCCG-1  2.303067   7.180564
## AAACCCACAGATGCGA-1    GBM AAACCCACAGATGCGA-1 -8.258726  -2.813064
## AAACCCACAGGTGAGT-1    GBM AAACCCACAGGTGAGT-1  8.892445  -7.549429
## AAACCCAGTCTTGCGG-1    GBM AAACCCAGTCTTGCGG-1  5.422224   5.287473
##                    seurat_clusters       cell_type umi_count log10_umi_count
## AAACCCAAGGCGATAC-1              10          t_cell      2226        3.347525
## AAACCCAAGGCTGTAG-1              12         myeloid      1063        3.026533
## AAACCCACAAGTCCCG-1               0       malignant     17883        4.252440
## AAACCCACAGATGCGA-1               7         myeloid      8172        3.912328
## AAACCCACAGGTGAGT-1               4 oligodendrocyte      9057        3.956984
## AAACCCAGTCTTGCGG-1               3       malignant      5612        3.749118
##                    percent_mt empty_drops_log_prob empty_drops_p_value
## AAACCCAAGGCGATAC-1   4.716981            -3881.078           9.999e-05
## AAACCCAAGGCTGTAG-1   2.916275            -2466.427           9.999e-05
## AAACCCACAAGTCCCG-1   1.509814           -24925.024           9.999e-05
## AAACCCACAGATGCGA-1   6.485560           -11318.990           9.999e-05
## AAACCCACAGGTGAGT-1   3.157779           -17035.760           9.999e-05
## AAACCCAGTCTTGCGG-1   0.427655           -13173.289           9.999e-05
##                    empty_drops_fdr nuclear_fraction_droplet_qc
## AAACCCAAGGCGATAC-1    0.0001044711                   0.1947243
## AAACCCAAGGCTGTAG-1    0.0001044711                   0.2766798
## AAACCCACAAGTCCCG-1    0.0000000000                   0.1843824
## AAACCCACAGATGCGA-1    0.0000000000                   0.2919902
## AAACCCACAGGTGAGT-1    0.0000000000                   0.3295617
## AAACCCAGTCTTGCGG-1    0.0000000000                   0.3795893
##                    nuclear_fraction_velocyto flag
## AAACCCAAGGCGATAC-1                 0.2612048 cell
## AAACCCAAGGCTGTAG-1                 0.3563123 cell
## AAACCCACAAGTCCCG-1                 0.2012495 cell
## AAACCCACAGATGCGA-1                 0.3429289 cell
## AAACCCACAGGTGAGT-1                 0.3589220 cell
## AAACCCAGTCTTGCGG-1                 0.3823998 cell</code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># How many cells in each dataset</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">qc_examples</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span></code></pre></div>
<pre><code>## 
##  GBM   HL   MB PBMC 
## 5516 2734 9658 9689</code></pre>
<p>We can look at the distribution of nuclear fraction scores directly:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">qc_examples</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">nuclear_fraction_droplet_qc</span><span class="op">)</span><span class="op">)</span>
<span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">sample</span><span class="op">)</span></code></pre></div>
<p><img src="dropletQC_files/figure-html/density-plot-1.png" width="480"></p>
<p>But it is often helpful to visualise the scores against the (log10) UMI count of each cell:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y<span class="op">=</span><span class="va">log10_umi_count</span><span class="op">)</span> <span class="op">+</span>
   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">sample</span><span class="op">)</span>
<span class="va">p</span></code></pre></div>
<p><img src="dropletQC_files/figure-html/against-umi-1.png" width="480"></p>
<p>Intuitively, empty droplets and damaged cells have a lower RNA content. We can see this more easily if we colour the cells by our classifications, saved in the ‘flag’ column:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>colour<span class="op">=</span><span class="va">flag</span><span class="op">)</span></code></pre></div>
<p><img src="dropletQC_files/figure-html/against-umi-labelled-1.png" width="624"></p>
<p>Droplets containing only ambient RNA will have a low nuclear fraction score, while damaged cells will tend to have a higher score. If your sample contains heterogeneous cell populations, it is recommended to assess these plots separately for each cell type. Your QC metrics are unlikley to be completely independent of the biology of your cells. For example, some cell types or states will naturally have a lower RNA content and may appear as outliers if not analysed separately. To illustrate this point we’ll have a quick look at the glioblastoma dataset, where a rough cell type annotation has already been performed:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Get the data for the glioblastoma example</span>
<span class="va">gbm</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">qc_examples</span>, <span class="va">sample</span> <span class="op">==</span> <span class="st">"GBM"</span><span class="op">)</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">gbm</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">umap_1</span>, y <span class="op">=</span> <span class="va">umap_2</span>, colour <span class="op">=</span> <span class="va">cell_type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="va">sample</span><span class="op">)</span></code></pre></div>
<p><img src="dropletQC_files/figure-html/gbm-umap-1.png" width="624"></p>
<p>Now we look at both the number of UMIs and the nuclear fraction score, split by cell type:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">&lt;-</span>
   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">gbm</span>,
          <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">nuclear_fraction_droplet_qc</span>, group <span class="op">=</span> <span class="va">cell_type</span>, fill <span class="op">=</span> <span class="va">cell_type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span>adjust <span class="op">=</span> <span class="fl">1.5</span>, alpha <span class="op">=</span> <span class="fl">.4</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Nuclear fraction score"</span><span class="op">)</span>
<span class="va">p2</span> <span class="op">&lt;-</span>
   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">gbm</span>,
          <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">log10_umi_count</span>, group <span class="op">=</span> <span class="va">cell_type</span>, fill <span class="op">=</span> <span class="va">cell_type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span>adjust <span class="op">=</span> <span class="fl">1.5</span>, alpha <span class="op">=</span> <span class="fl">.4</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"log10(UMI count)"</span><span class="op">)</span>
<span class="va">p1</span></code></pre></div>
<p><img src="dropletQC_files/figure-html/qc-metrics-celltype-1.png" width="624"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p2</span></code></pre></div>
<p><img src="dropletQC_files/figure-html/qc-metrics-celltype-2.png" width="624"></p>
<p>The nuclear fraction scores are approximately normally distributed for each cell type. A value of +/- three standard deviations from the mean is a reasonable, if conservative, starting point for assigning thresholds. Here is an example of applying thresholds to the glioblastoma dataset:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Define nuclear fraction thresholds</span>
<span class="va">thresholds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tribble.html">tribble</a></span><span class="op">(</span>
   <span class="op">~</span><span class="va">cell_type</span>,  <span class="op">~</span><span class="va">lower</span>,  <span class="op">~</span><span class="va">upper</span>,
   <span class="st">"endothelial"</span>, <span class="fl">0.05</span>, <span class="fl">0.49</span>,
   <span class="st">"malignant"</span>, <span class="fl">0.05</span>, <span class="fl">0.59</span>,
   <span class="st">"myeloid"</span>, <span class="fl">0.06</span>, <span class="fl">0.56</span>,
   <span class="st">"oligodendrocyte"</span>, <span class="fl">0.09</span>, <span class="fl">0.78</span>,
   <span class="st">"t_cell"</span>, <span class="fl">0.12</span>, <span class="fl">0.67</span><span class="op">)</span>

<span class="co"># Plot thresholds</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">gbm</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">nuclear_fraction_droplet_qc</span>, y<span class="op">=</span><span class="va">log10_umi_count</span>, colour<span class="op">=</span><span class="va">flag</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.title<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">cell_type</span><span class="op">)</span><span class="op">+</span>
   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">thresholds</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="va">lower</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span>
   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">thresholds</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="va">upper</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span></code></pre></div>
<p><img src="dropletQC_files/figure-html/apply-thresholds-1.png" width="720"> Using a combination of metrics, such as mitchondrial gene content, number of UMIs and the nuclear fraction score can often be helpful.</p>
<p>It will usually be desirable to exclude empty droplets from your dataset. However, depending on the nature of your sample and your biological question, damaged cells may still retain plenty of useful information. In that case you may want to just mark these cells, keeping the metadata at hand throughout downstream analyses.</p>
<p>It’s difficult to provide guidance on where to place thresholds that will be applicable in every circumstance. However, before deciding to exclude a population of cells we recommend pausing to consider whether the data are consistent with your expectations and knowledge of the biology of each cell type in your sample:</p>
<ul>
<li><p>Does it make sense that a cluster of large malignant cells has a very low RNA content, or are they more likely to represent empty droplets?</p></li>
<li><p>Should a cluster of t-cells which appears to be an outlier be excluded or do they just have a lower average RNA content?</p></li>
<li><p>Do the sensitive cell types in a brain sample, astrocytes and neurons, represent the majority of damaged cells?</p></li>
</ul>
<p>Thinking through questions like these will be more fruitful than blindly following rules of thumb that may be inappropriate for your dataset or research question.</p>
</div>
<div id="gene-level-counts" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-level-counts" class="anchor"></a>Gene-level counts</h2>
<p>Note that if you have run a counting pipeline, such as <a href="http://velocyto.org/">velocyto</a> to quantify spliced and unspliced counts, you already have the information you need to calculate the nuclear fraction. If for example you have run <code>velocyto run10x</code> you could use the following:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Import loom file</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> sc.read_loom(<span class="st">"velocyto.loom"</span>)</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the nuclear fraction using the spliced and unspliced matrices</span></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>exon_sum <span class="op">=</span> adata.layers[<span class="st">'spliced'</span>].<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>intron_sum <span class="op">=</span> adata.layers[<span class="st">'unspliced'</span>].<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>nuclear_fraction <span class="op">=</span> intron_sum<span class="op">/</span>(exon_sum <span class="op">+</span> intron_sum)</span></code></pre></div>
<p>Counting pipelines like these are much more powerful, as they provide spliced and unspliced counts <strong>per gene</strong> as well as per cell barcode. This additional information allows you to perform further analyses, such as RNA velocity, that can take advantage of this additional information. The trade-off here is time - <code><a href="../reference/nuclear_fraction_tags.html">dropletQC::nuclear_fraction_tags</a></code> ignores UMIs and gene-level information in order to calculate the nuclear fraction score quickly. Here is a comparison of the nuclear fraction scores calculated with <code>dropletQC::nuclear_fraction</code> <em>vs</em> the output from <code>velocyto</code>:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">qc_examples</span>,
       <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">nuclear_fraction_droplet_qc</span>, y <span class="op">=</span> <span class="va">nuclear_fraction_velocyto</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span> <span class="op">~</span> <span class="va">sample</span><span class="op">)</span></code></pre></div>
<p><img src="dropletQC_files/figure-html/velocyto-dropletqc-comparison-1.png" width="480"></p>
<p>As always, your choice of tool will depend on what you are trying to do. If you just want to calculate a nuclear fraction score for QC, this package should be sufficient. If you’re interested in looking at gene expression dynamics, a more sophisticated tool like <code>velocyto</code> will be required.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Walter Muskovic.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
